components/tikv_util/src/config.rs

```diff
@@ -1456,68 +1433,88 @@ impl RaftDataStateMachine {
         Ok(())
     }
 
-    /// Enters the `Migrating` state and returns the source directory if a
-    /// migration is needed. Otherwise prepares the target directory for
-    /// opening.
-    pub fn before_open_target(&mut self) -> Option<PathBuf> {
+    /// Returns whether a migration is needed. When it's needed, enters the `Migrating`
+    /// state. Otherwise prepares the target directory for opening.
+    pub fn before_open_target(&mut self) -> bool {
         // Clean up trash directory if there is any.
-        for p in [
-            &self.source,
-            &self.source_staging,
-            &self.target,
-            &self.target_staging,
-        ] {
+        for p in [&self.source, &self.target] {
             let trash = p.with_extension("REMOVE");
             if trash.exists() {
                 fs::remove_dir_all(&trash).unwrap();
             }
         }
-        if Self::data_exists(&self.target_staging) {
-            // InvMigrating -> Completed
-            assert!(!Self::data_exists(&self.source_staging));
-            Self::remove_dir_safe(&self.source);
-            Self::rename_dir_safe(&self.target_staging, &self.target);
-            return None;
-        }
-        if Self::data_exists(&self.source) {
-            // Init -> Migrating
-            assert!(!Self::data_exists(&self.source_staging));
-            Self::rename_dir_safe(&self.source, &self.source_staging);
-        } else if !Self::data_exists(&self.source_staging) {
-            // No source data.
-            return None;
+        if !Self::data_exists(&self.source) {
+            // Recover from Completed state.
+            if self.in_progress_marker.exists() {
+                Self::must_remove(&self.in_progress_marker);
+            }
+            return false;
+        } else if self.in_progress_marker.exists() {
+            // Recover from Migrating state.
+            let real_source = PathBuf::from(fs::read_to_string(&self.in_progress_marker).unwrap());
+            if real_source == self.target {
+                Self::must_remove(&self.source);
+                return false;
+            } else {
+                Self::must_remove(&self.target);
+            }
+        } else {
+            // Init -> Migrating.
+            fs::write(&self.in_progress_marker, self.source.to_str().unwrap()).unwrap();
```

Comment:
What if it crashes here and leave the content of the file truncated?

Reply:
Ok, I'll add a double slash at the end as an EOF marker.

Reply:
Then you also need to sync the file.